Wrap lines based on word boundries

gfontenot opened the pull request:
This makes line wrapping a little smarter. Instead of hard-wrapping on
80 (or however many) characters, we can wrap based on word boundries if
adding another word to the line pushes it past 80 (or however many)
characters.

This also makes the line length arbitrary.

gfontenot committed:
Wrap lines based on word boundries

This makes line wrapping a little smarter. Instead of hard-wrapping on
80 (or however many) characters, we can wrap based on word boundries if
adding another word to the line pushes it past 80 (or however many)
characters.

This also makes the line length arbitrary.

gfontenot committed:
Remove unnecessary functions

We're splitting on word boundries now, so we actually don't need to
worry about empty strings or trailing/leading whitespace. Nice.

gabebw commented on Main.hs:
```diff
@@ -1,12 +1,15 @@
-trim :: String -> String
-trim = T.unpack . T.strip . T.pack
+lineIsTooLong :: Int -> String -> Bool
+lineIsTooLong n xs = case length xs `compare` n of
```
Can this be `length xs > n`?

gfontenot replied:
... yep

could have sworn I tried that.

gabebw commented on src/Main.hs:
```diff
@@ -1,12 +1,15 @@
+splitWordsAt :: Int -> [String] -> String -> [String]
+splitWordsAt _ [] x = [x]
+splitWordsAt n xs x
+  | lineIsTooLong n (unwords [last xs, x]) = xs ++ [x]
```
Can you explain this a little mroe? The `unwords [last xs, x]` is hard to read.

gabebw commented on src/Main.hs:
```diff
@@ -1,12 +1,15 @@
-trim :: String -> String
-trim = T.unpack . T.strip . T.pack
+lineIsTooLong :: Int -> String -> Bool
+lineIsTooLong n xs = case length xs `compare` n of
```
Can this be `length xs > n`?

gfontenot commented:
@gabebw To explain the intent behind this, we split the line up into words, then iterate through those words to create new lines. For each iteration:

 - We take the last line, which is the line currently being constructed (`last xs`)
 - We check to see what the new line length would be if we added the current word to it
 - If the line is over the specified length, we create a new line by appending the current word in an empty list
 - If the line is still shorter (or equal to) the specified line length, we append the current word to the current line (`unwords [last xs, x]`)

I feel like there's a refactor (or multiple) that could be done here to help with clarity, but I'm having trouble seeing them.

gabebw commented on src/Main.hs:
```diff
@@ -1,12 +1,13 @@
-wrap :: String -> [String]
-wrap = map trim . combine . splitAt 80
+wrap :: Int -> String -> [String]
+wrap n s = foldl (splitWordsAt n) [] . words $ s
```
I'd probably write the last bit of this as `$ words s` instead of `. words $ s`. Or do `.words` and leave off the `s`! Point freeeee

gfontenot replied:
This _was_ point free and I was super happy about it, but then that pesky `n` showed up and screwed everything up. Good call re: `$` though. Makes much more sense that way.

gfontenot committed:
PR comments

gfontenot commented:
sure would be great someone cool like @patbrisbin would help give me feedback too

man

that'd be swell

pbrisbin commented:
Wow, give a guy a chance. You mention me for review and immediately merge? That's not our PR process!

pbrisbin commented:
I might've tried a recursive approach: figure out the first line, then `wrap` the rest:

```hs
import Data.List

wrap :: Int -> String -> [String]
wrap _ [] = []
wrap n s = firstLine : wrap n rest
  where
    -- +1 because splitAt returns the delimiter in rest
    (firstLine, rest) = splitAt (bestSpace + 1) s

    -- returns the length of s if no space is found
    bestSpace = head . (++ [length s]) $
        dropWhile (<= n) $ findIndices (== ' ') s
```

Note: may not handle edge cases like double-space or non-space whitespace.

gfontenot commented:
Can't stop the progress here at git hub dot com slash gee fontenot, Pat
